pkg/restore/restore.go

```diff
@@ -1468,3 +1471,140 @@ func isCompleted(obj *unstructured.Unstructured, groupResource schema.GroupResou
+	// resource twice even if it's in the ordered list twice.
+	for _, resource := range getOrderedResources(ctx.resourcePriorities, backupResources) {
+		// try to resolve the resource via discovery to a complete group/version/resource
+		gvr, _, err := ctx.discoveryHelper.ResourceFor(schema.ParseGroupResource(resource).WithVersion(""))
+		if err != nil {
+			ctx.log.WithField("resource", resource).Infof("Skipping restore of resource because it cannot be resolved via discovery")
+			continue
+		}
+		groupResource := gvr.GroupResource()
+
+		// check if we've already restored this resource (this would happen if the resource
+		// we're currently looking at was already restored because it was a prioritized
+		// resource, and now we're looking at it as part of the backup contents).
+		if processedResources.Has(groupResource.String()) {
+			ctx.log.WithField("resource", groupResource.String()).Debugf("Skipping restore of resource because it's already been processed")
+			continue
+		}
+
+		// check if the resource should be restored according to the resource includes/excludes
+		if !ctx.resourceIncludesExcludes.ShouldInclude(groupResource.String()) {
+			ctx.log.WithField("resource", groupResource.String()).Infof("Skipping restore of resource because the restore spec excludes it")
+			continue
+		}
+
+		// we don't want to explicitly restore namespace API objs because we'll handle
+		// them as a special case prior to restoring anything into them
+		if groupResource == kuberesource.Namespaces {
+			continue
+		}
+
+		// check if the resource is present in the backup
+		resourceList := backupResources[groupResource.String()]
+		if resourceList == nil {
+			ctx.log.WithField("resource", groupResource.String()).Debugf("Skipping restore of resource because it's not present in the backup tarball")
+			continue
+		}
+
+		// iterate through each namespace that contains instances of the resource and
+		// restore them
+		for namespace, items := range resourceList.ItemsByNamespace {
+			if namespace != "" && !ctx.namespaceIncludesExcludes.ShouldInclude(namespace) {
+				ctx.log.Infof("Skipping namespace %s", namespace)
+				continue
+			}
+
+			// get target namespace to restore into, if different
+			// from source namespace
+			targetNamespace := namespace
+			if target, ok := ctx.restore.Spec.NamespaceMapping[namespace]; ok {
+				targetNamespace = target
+			}
+
+			res, w, e := ctx.getSelectedRestoreableItems(groupResource.String(), targetNamespace, namespace, items)
+			restoreResourceCollection = append(restoreResourceCollection, res)
+
+			warnings.Merge(&w)
+			errs.Merge(&e)
+		}
+
+		// record that we've restored the resource
+		processedResources.Insert(groupResource.String())
+	}
+	return
+}
+
+func (ctx *restoreContext) getSelectedRestoreableItems(resource, targetNamespace, originalNamespace string, items []string) (res restoreableResource, warnings Result, errs Result) {
```

Comment:
If I'm reading this code correctly, `getOrderedResourceCollection` is very similar to the previous restore loop in that it's applying the selection criteria like namespace, and includes/excludes. However, the change here is that this function, `getSelectedRestoreableItems` then groups those resources by namespace and applies the Kubernetes label selection. It replaces some of the functionality of the `restoreResource` function, which should be removed since it's no longer relevant with this PR, from what I can tell.

Do I have that right? If so please add a comment to clarify that's what this function is doing, so we can keep track of it.

Also, based on this code it looks like we're unmarshaling twice.

Reply:
@nrb that is correct. We are unmarshling twice. Unmarshling is needed at first to apply selection criteria, and then finally to actually restore the object. Another alternative was to keep entire object in-memory until it's restored which isnt very attractive.

Reply:
and your'e correct about the duplicate code from restoreResource function. thanks for pointing that out. will clean it up

Reply:
> Another alternative was to keep entire object in-memory until it's restored which isnt very attractive.

Right, that is the trade we'd have to make. I honestly think we will end up doing this at some point, though, as we'd like to form a graph of objects. For now I think this will work and we can revisit the unmarshalling mechanisms in the future.